server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4252,39 +4257,113 @@ public void testCleanupCommitsWhenReleaseSnapshot() throws Exception {
             final Engine.IndexResult result = engine.index(replicaIndexForDoc(doc, 1L, id, false));
             assertThat(result.isCreated(), equalTo(false));
         }
         SegmentInfos lastCommitInfo = engine.getLastCommittedSegmentInfos();
-        assertThat(engine.getTranslog().uncommittedOperations(), equalTo(numDocs));
+        assertThat(translog.totalOperationsByMinGen(translog.uncommittedGeneration()), equalTo(numDocs));
         assertThat(engine.shouldPeriodicallyFlush(), equalTo(true));
         engine.flush(false, false);
         assertThat(engine.getLastCommittedSegmentInfos(), not(sameInstance(lastCommitInfo)));
-        assertThat(engine.getTranslog().uncommittedOperations(), equalTo(0));
+        assertThat(translog.totalOperationsByMinGen(translog.uncommittedGeneration()), equalTo(0));
+
+        // If the new index commit still points to the same translog generation as the current index commit,
+        // we should not enable the periodically flush condition; otherwise we can get into an infinite loop of flushes.
+        engine.getLocalCheckpointTracker().generateSeqNo(); // create a gap here
+        for (int id = 0; id < numDocs; id++) {
+            if (randomBoolean()){
+                translog.rollGeneration();
+            }
+            final ParsedDocument doc = testParsedDocument("new" + id, null, testDocumentWithTextField(), SOURCE, null);
+            long seqno = engine.getLocalCheckpointTracker().generateSeqNo();
+            final Engine.IndexResult result = engine.index(replicaIndexForDoc(doc, 2L, seqno, false));
+            assertThat(result.isCreated(), equalTo(true));
+        }
+        lastCommitInfo = engine.getLastCommittedSegmentInfos();
+        if (engine.shouldPeriodicallyFlush()) {
+            engine.flush();
+            assertThat(engine.getLastCommittedSegmentInfos(), not(sameInstance(lastCommitInfo)));
+        }
+        assertThat(engine.shouldPeriodicallyFlush(), equalTo(false));
     }
 
+    public void testStressShouldPeriodicallyFlush() throws Exception {
+        final long flushThreshold = randomLongBetween(100, 5000);
+        final long generationThreshold = randomLongBetween(1000, 5000);
+        final IndexSettings indexSettings = engine.config().getIndexSettings();
+        final IndexMetaData indexMetaData = IndexMetaData.builder(indexSettings.getIndexMetaData())
+            .settings(Settings.builder().put(indexSettings.getSettings())
+                .put(IndexSettings.INDEX_TRANSLOG_GENERATION_THRESHOLD_SIZE_SETTING.getKey(), generationThreshold + "b")
+                .put(IndexSettings.INDEX_TRANSLOG_FLUSH_THRESHOLD_SIZE_SETTING.getKey(), flushThreshold + "b")).build();
+        indexSettings.updateIndexMetaData(indexMetaData);
+        engine.onSettingsChanged();
+        final int iterations = scaledRandomIntBetween(100, 1000);
+        final List<Long> pendingSeqNo = new ArrayList<>();
```

Comment:
maybe instead of pendingSeqNo, randomg take a seq# from the range (localCheckpoint:localCheckpoint+5] ?

Reply:
Boaz, thanks for this hint. This change makes the test failed more than 50% without the patch and also fixed the file descriptor issue.